src/main/java/com/puppycrawl/tools/checkstyle/filters/PatchXpathFilterElement.java

```diff
@@ -73,26 +78,99 @@ private boolean isFileNameMatching(TreeWalkerAuditEvent event) {
+        return result;
+    }
+
+    private boolean lineMatching(int currentLine) {
+        boolean result = false;
+        for (List<Integer> singleLineRangeList : lineRangeList) {
+            final int startLine = singleLineRangeList.get(0) + 1;
+            final int endLine = singleLineRangeList.get(1) + 1;
+            result = currentLine >= startLine && currentLine < endLine;
+            if (result) {
+                break;
             }
         }
         return result;
     }
 
     /**
-     * Is matching by xpath query.
+     * Is matching by context strategy.
      *
      * @param event event
      * @return true if it is matching or not set.
      */
-    private boolean isXpathQueryMatching(TreeWalkerAuditEvent event) {
-        return true;
+    private boolean isContextStrategyCheck(TreeWalkerAuditEvent event) {
+        boolean result = false;
+        final String checkShortName = getCheckShortName(event);
+        if ("OuterTypeNumberCheck".equals(checkShortName)) {
+            result = true;
+        }
+        else if ("ReturnCountCheck".equals(checkShortName)
+                || "MethodCountCheck".equals(checkShortName)
+                || "AnonInnerLengthCheck".equals(checkShortName)
+                || "ExecutableStatementCountCheck".equals(checkShortName)
+                || "MethodLengthCheck".equals(checkShortName)) {
+            final DetailAST eventAst = getEventAst(event);
+            final Set<Integer> childAstLineNoList = getChildAstLineNo(eventAst, event);
+            for (Integer currentLine : childAstLineNoList) {
+                result = lineMatching(currentLine);
+                if (result) {
+                    break;
+                }
+            }
+        }
+        return result;
+    }
+
+    private String getCheckShortName(TreeWalkerAuditEvent event) {
+        final String[] checkNames = event.getLocalizedMessage().getSourceName().split("\\.");
+        return checkNames[checkNames.length - 1];
+    }
+
+    private DetailAST getEventAst(TreeWalkerAuditEvent event) {
+        DetailAST curNode = event.getRootAst();
+        DetailAST eventAst = null;
+        while (curNode != null) {
+            if (isMatchingAst(curNode, event)) {
+                eventAst = curNode;
+                break;
+            }
+            DetailAST toVisit = curNode.getFirstChild();
+            while (curNode != null && toVisit == null) {
+                toVisit = curNode.getNextSibling();
+                curNode = curNode.getParent();
+            }
+            curNode = toVisit;
+        }
+        return eventAst;
+    }
+
+    private Set<Integer> getChildAstLineNo(DetailAST ast, TreeWalkerAuditEvent event) {
+        final Set<Integer> childAstLineNoSet = new HashSet<>();
+        DetailAST curNode = ast;
+        while (curNode != null) {
+            DetailAST toVisit = curNode.getFirstChild();
+            if (toVisit != null) {
+                childAstLineNoSet.add(toVisit.getLineNo());
```

Comment:
This will collect all line numbers for all child nodes. Basically we only need start and end for `ast`.
`start` is line number of `ast` itself and in this method you need to find max line number among `ast`'s children which will be `end`.

Reply:
@baratali according to `lineMatching` function, we need all line numbers.

Reply:
well, we need one more lineMatching function then :)
something like this:
```java
    private boolean lineMatching(int astLineStart, int astLineEnd) {
        boolean result = false;
        for (List<Integer> singleLineRangeList : lineRangeList) {
            final int startLine = singleLineRangeList.get(0) + 1;
            final int endLine = singleLineRangeList.get(1) + 1;
            result = astLineStart < endLine && astLineEnd >= startLine;
            if (result) {
                break;
            }
        }
        return result;
    }
```

Reply:
sorry, I am not clear about this code: ` result = astLineStart < endLine && astLineEnd >= startLine;`

Reply:
@HuGanghui , ok, I created new issue: https://github.com/checkstyle/patch-filters/issues/138